plugins/pixart-rf/fu-pxi-ble-device.c

```diff
@@ -210,9 +216,23 @@ fu_pxi_ble_device_search_hid_usage_page(guint8 *report_descriptor,
 
 		memmove(usage_page_tmp, &report_descriptor[pos + 1], report_size);
 		if (memcmp(usage_page, usage_page_tmp, usage_page_sz) == 0) {
+			gint index = 0;
 			g_debug("hit item: %x  ", item);
 			fu_dump_raw(G_LOG_DOMAIN, "usage_page", usage_page, report_size);
 			g_debug("hit pos %d", pos);
+			index = pos;
+
+			while (index <= size) {
+				if(report_descriptor[index] == 0x85) {
+					index = index + 1;
+					*report_id = report_descriptor[index];
+					g_debug("repor_id: %x  ", *report_id);
+					break;
+				}
+
+				index++;
+			}
+
```

Comment:
are you trying to go from usage page to report ID? I merged some code for this very recently: https://github.com/fwupd/fwupd/commit/fca53b31d25a9570de4d5c279e4a023ca1dd42ab

Reply:
It's useful!!  I replace my code with it

Reply:
Yell if you need any sample code.

Reply:
After meet with our OEM customer, they decided to use this version. In the next step, we will integrate the new official HID related API into our plugin

Reply:
Thanks your advising.

Reply:
> After meet with our OEM customer, they decided to use this version

Right, but your OEM customer isn't the one maintaining this code for the next 10 years. I'd like you to use the FuHidDescriptor code before I will approve this PR.

Reply:
Sure! The FuHidDescriptor[1] object subclasess FuFirmware so it's a few lines of code: 

	g_autoptr(GBytes) fw = g_bytes_new(pixart_buffer, pixart_sizeofbuffer);
	g_autoptr(FuHidDescriptor) descriptor = fu_hid_descriptor_new();
	g_autoptr(FuHidReport) report = NULL;
	g_autoptr(FuFirmware) item_id = NULL;

	if (!fu_firmware_parse(descriptor, fw, FWUPD_INSTALL_FLAG_NONE, error))
		return FALSE;
	report = fu_hid_descriptor_find_report_by_usage(FU_HID_DESCRIPTOR(descriptor),
							pixart_usage_page, 
							pixart_usage,
							error);
	if (report == NULL)
		return FALSE;
	item_id = fu_firmware_get_image_by_id(FU_FIRMWARE(report), "report-id", error);
	if (item_id == NULL)
		return FALSE;
	pixart_report_id = fu_hid_report_item_get_value(FU_HID_REPORT_ITEM(item_id));

[1] https://github.com/fwupd/fwupd/blob/main/libfwupdplugin/fu-hid-descriptor.h

Reply:
0x06, 0x02, 0xFF,  // Usage Page (Vendor Defined 0xFF02)
    0x09, 0x01,        // Usage (0x01)
    0xA1, 0x01,        // Collection (Application)
    0x75, 0x08,        //   Report Size (8)
    0x95, 0x14,        //   Report Count (20)
    0x85, 0x09,        //   Report ID (9)
    0x09, 0x01,        //   Usage (0x01)
    0xB1, 0x02,        //   Feature (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position,Non-volatile)
    0xC0,              // End Collection

Reply:
guint8 pixart_buffer[512] = {0x0};
	g_autoptr(GBytes) fw = g_bytes_new(rpt_desc.value, rpt_desc.size);
	g_autoptr(FuHidDescriptor) descriptor = fu_hid_descriptor_new();
	g_autoptr(FuHidReport) report = NULL;
	g_autoptr(FuFirmware) item_id = NULL;

	if (!fu_firmware_parse(descriptor, fw, FWUPD_INSTALL_FLAG_NONE, error))
		return FALSE;
		
	report = fu_hid_descriptor_find_report_by_usage(FU_HID_DESCRIPTOR(descriptor),
							0xff02, 
							0x01,
							error);
	
	if (report == NULL)
		return FALSE;
	item_id = fu_firmware_get_image_by_id(FU_FIRMWARE(report), "report-id", error);
	if (item_id == NULL)
		return FALSE;
	guint8 pixart_report_id = fu_hid_report_item_get_value(FU_HID_REPORT_ITEM(item_id));
	g_debug("#pixart_report_id: %x\r\n",pixart_report_id);

Reply:
the parsing result the report id is 8, but the correct id is 9, Is any condition I miss ?

Reply:
<pre><font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">#pixart_report_id: 8</font>

<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">target usage_page:01 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit item: 6  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     01 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 285</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 8  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">target usage_page:02 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit item: 6  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     02 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 303</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 9  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">target usage_page:00 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     00 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 206</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 5  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">usage_page:     00 ff </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">hit pos 257</font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">repor_id: 7  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">report_type: 2  </font>
<font color="#4E9A06">09:22:18.144 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     08 28 </font>
<font color="#4E9A06">09:22:18.477 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     09 10 </font>
<font color="#4E9A06">09:22:18.491 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     09 23 </font>
<font color="#4E9A06">09:22:18.529 FuPluginPixartRf     </font><font color="#3465A4">GetFeature:     0e 09 23 00 33 2e 30 2e 32 84 d8 </font>
<font color="#4E9A06">09:22:18.529 FuPluginPixartRf     </font><font color="#3465A4">doing fixup for old bluez version</font>
<font color="#4E9A06">09:22:18.529 FuPluginPixartRf     </font><font color="#3465A4">SetFeature:     09 2b </font>
<font color="#4E9A06">09:22:18.574 FuPluginPixartRf     </font><font color="#3465A4">GetFeature:</font>
</pre>

Reply:
one sec; you've found a bug in the descriptor parser. I've fixed the bug, but give me an hour to write some tests.

	@@ -260,6 +265,11 @@ fu_hid_descriptor_find_report_by_usage(FuHidDescriptor *self,
					continue;
			}
	 
	+               /* don't return the top-level collection */
	+               item_id = fu_firmware_get_image_by_idx(FU_FIRMWARE(report), FU_HID_ITEM_TAG_COLLECTION, NULL);
	+               if (item_id != NULL)
	+                       continue;
	+

Reply:
Okay, lets fix the API then :) Give me a few mins.

Reply:
@jyu99 does something like this work? https://github.com/fwupd/fwupd/pull/6015 -- you can then do something like:

	report = fu_hid_descriptor_find_report(FU_HID_DESCRIPTOR(firmware),
					       &error,
					       "usage-page",
					       0xFF02,
					       "usage",
					       0x01,
					       "feature",
					       0x02,
					       NULL);

Reply:
It's fine.

Reply:
please tell me if you complete the HID API fix,and then I will recreate the pull request and update my code.Thanks!

Reply:
Okay, merged -- please cherry pick that commit into your branch or rebase on top of master.